File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/publisher/acknowledge/PublisherAcknowledgeSpec.java
package io.micronaut.rabbitmq.docs.publisher.acknowledge;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
class PublisherAcknowledgeSpec {

    @Test
    void testPublisherAcknowledgement(ProductClient productClient) {

        AtomicInteger successCount = new AtomicInteger(0);
        AtomicInteger errorCount = new AtomicInteger(0);

// tag::producer[]
Publisher<Void> publisher = productClient.sendPublisher("publisher body".getBytes());
CompletableFuture<Void> future = productClient.sendFuture("future body".getBytes());

Subscriber<Void> subscriber = new Subscriber<Void>() {
    @Override
    public void onSubscribe(Subscription subscription) { }

    @Override
    public void onNext(Void aVoid) {
        throw new UnsupportedOperationException("Should never be called");
    }

    @Override
    public void onError(Throwable throwable) {
        // if an error occurs
        errorCount.incrementAndGet();
    }

    @Override
    public void onComplete() {
        // if the publish was acknowledged
        successCount.incrementAndGet();
    }
};
publisher.subscribe(subscriber);
future.whenComplete((v, t) -> {
    if (t == null) {
        successCount.incrementAndGet();
    } else {
        errorCount.incrementAndGet();
    }
});
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                errorCount.get() == 0 &&
                successCount.get() == 2
        );
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/rabbitmq/docs/publisher/acknowledge/PublisherAcknowledgeSpec.kt
package io.micronaut.rabbitmq.docs.publisher.acknowledge;

import io.kotest.assertions.nondeterministic.eventually
import io.kotest.core.spec.style.AnnotationSpec
import io.kotest.matchers.shouldBe
import io.micronaut.context.annotation.Property
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import io.micronaut.test.support.TestPropertyProvider
import jakarta.inject.Inject
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.Duration.Companion.seconds
import io.micronaut.rabbitmq.testcontainers.RabbitMQ

@MicronautTest
@Property(name = "spec.name", value = "PublisherAcknowledgeSpec")
class PublisherAcknowledgeSpec : TestPropertyProvider, AnnotationSpec() {

    override fun getProperties(): Map<String, String> {
        return RabbitMQ.getProperties()
    }

    @Inject
    lateinit var productClient: ProductClient

    @Test
    suspend fun testPublisherAcknowledgement() = coroutineScope {
        val successCount = AtomicInteger(0)
        val errorCount = AtomicInteger(0)

        // tag::producer[]
        val publisher = productClient.sendPublisher("publisher body".toByteArray())
        val future = productClient.sendFuture("future body".toByteArray())
        val deferred = async {
            productClient.sendSuspend("suspend body".toByteArray())
        }

        val subscriber = (object: Subscriber<Void> {
            override fun onSubscribe(subscription: Subscription) { }

            override fun onNext(aVoid: Void) {
                throw UnsupportedOperationException("Should never be called")
            }

            override fun onError(throwable: Throwable) {
                // if an error occurs
                errorCount.incrementAndGet()
            }

            override fun onComplete() {
                // if the publish was acknowledged
                successCount.incrementAndGet()
            }
        })
        publisher.subscribe(subscriber)
        future.handle { _, t ->
            if (t == null) {
                successCount.incrementAndGet()
            } else {
                errorCount.incrementAndGet()
            }
        }
        deferred.invokeOnCompletion {
            if (it == null) {
                successCount.incrementAndGet()
            } else {
                errorCount.incrementAndGet()
            }
        }
// end::producer[]

        eventually(10.seconds) {
            errorCount.get() shouldBe 0
            successCount.get() shouldBe 3
        }
    }

}
